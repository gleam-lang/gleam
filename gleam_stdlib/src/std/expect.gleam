// TODO: Move this module into another package so it can be used as a
// dep only in test.

pub external type Expectation;

pub external fn equal(a, a) -> Expectation = "gleam__stdlib" "expect_equal";

pub external fn not_equal(a, a) -> Expectation = "gleam__stdlib" "expect_not_equal";

pub external fn true(Bool) -> Expectation = "gleam__stdlib" "expect_true";

pub external fn false(Bool) -> Expectation = "gleam__stdlib" "expect_false";

pub external fn is_ok(Result(a, b)) -> Expectation = "gleam__stdlib" "expect_is_ok";

pub external fn is_error(Result(a, b)) -> Expectation = "gleam__stdlib" "expect_is_error";

pub fn fail() {
  true(False)
}
