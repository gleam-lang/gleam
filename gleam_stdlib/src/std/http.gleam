// HTTP standard method as defined by RFC 2616, and PATCH which is defined by
// RFC 5789.
//
pub enum Method =
  | Get
  | Post
  | Head
  | Put
  | Delete
  | Trace
  | Connect
  | Options
  | Patch
