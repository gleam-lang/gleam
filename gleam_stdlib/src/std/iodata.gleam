pub external type Iodata;

pub external fn prepend(Iodata, String) -> Iodata =
  "gleam__stdlib" "iodata_prepend";

pub external fn append(Iodata, String) -> Iodata =
  "gleam__stdlib" "iodata_append";

pub external fn prepend_iodata(Iodata, Iodata) -> Iodata =
  "gleam__stdlib" "iodata_prepend";

pub external fn append_iodata(Iodata, Iodata) -> Iodata =
  "gleam__stdlib" "iodata_append";

pub external fn from_strings(List(String)) -> Iodata =
  "gleam__stdlib" "identity";

pub external fn concat(List(Iodata)) -> Iodata =
  "gleam__stdlib" "identity";

pub external fn new(String) -> Iodata =
  "gleam__stdlib" "identity";

pub external fn to_string(Iodata) -> String =
  "erlang" "iolist_to_binary";

pub external fn byte_size(Iodata) -> Int =
  "erlang" "iolist_size";

pub external fn from_float(Float) -> Iodata =
  "io_lib_format" "fwrite_g";

pub external fn lowercase(Iodata) -> Iodata = "string" "lowercase"

pub external fn uppercase(Iodata) -> Iodata = "string" "uppercase"

pub external fn reverse(Iodata) -> Iodata = "string" "reverse"

enum Direction =
  | All

external fn erl_split(Iodata, String, Direction) -> List(Iodata) =
  "string" "split"

pub fn split(iodata, on) {
  erl_split(iodata, on, All)
}

external fn erl_replace(Iodata, String, String, Direction) -> Iodata =
  "string" "replace"

pub fn replace(iodata, pattern, replacement) {
  erl_replace(iodata, pattern, replacement, All)
}

pub external fn is_equal(Iodata, Iodata) -> Bool = "string" "equal"

pub external fn is_empty(Iodata) -> Bool = "string" "is_empty"
