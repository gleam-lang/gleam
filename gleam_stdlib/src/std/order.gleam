pub enum Order =
  | Lt
  | Eq
  | Gt
;

pub fn reverse(order) {
  case order {
  | Lt -> Gt
  | Eq -> Eq
  | Gt -> Lt
  }
}

pub fn to_int(order) {
  case order {
  | Lt -> -1
  | Eq -> 0
  | Gt -> 1
  }
}

pub fn compare(a, b) {
  case {a, b} {
  | {Lt, Lt} -> Eq
  | {Lt, _} -> Lt
  | {Eq, Eq} -> Eq
  | {Gt, Gt} -> Eq
  | {Eq, Gt} -> Lt
  | _ -> Gt
  }
}

pub fn max(a, b) {
  case {a, b} {
  | {Gt, _} -> Gt
  | {Eq, Lt} -> Eq
  | _ -> b
  }
}

pub fn min(a, b) {
  case {a, b} {
  | {Lt, _} -> Lt
  | {Eq, Gt} -> Eq
  | _ -> b
  }
}
