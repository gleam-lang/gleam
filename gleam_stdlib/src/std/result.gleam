// Result represents the result of something that may succeed or fail.
// `Ok` means it was successful, `Error` means it failed.

pub fn is_ok(result) {
  case result {
  | Error(_) -> False
  | Ok(_) -> True
  }
}

pub fn is_error(result) {
  case result {
  | Ok(_) -> False
  | Error(_) -> True
  }
}

pub fn map(result, fun) {
  case result {
  | Ok(x) -> Ok(fun(x))
  | Error(e) -> Error(e)
  }
}

pub fn map_error(result, fun) {
  case result {
  | Ok(_) -> result
  | Error(error) -> Error(fun(error))
  }
}

pub fn flatten(result) {
  case result {
  | Ok(x) -> x
  | Error(error) -> Error(error)
  }
}

pub fn then(result, fun) {
  case result {
  | Ok(x) -> fun(x)
  | Error(e) -> Error(e)
  }
}

pub fn unwrap(result, default) {
  case result {
  | Ok(v) -> v
  | Error(_) -> default
  }
}
