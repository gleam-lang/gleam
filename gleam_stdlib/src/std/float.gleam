import std/iodata

pub enum NotAFloat =
  | NotAFloat

pub external fn parse(String) -> Result(Float, NotAFloat) =
  "gleam__stdlib" "parse_float";

pub fn to_string(f) {
  f
  |> iodata:from_float
  |> iodata:to_string
}

pub external fn ceiling(Float) -> Float = "math" "ceil";

pub external fn floor(Float) -> Float = "math" "floor";

pub external fn round(Float) -> Int = "erlang" "round";

pub external fn truncate(Float) -> Int = "erlang" "trunc";
