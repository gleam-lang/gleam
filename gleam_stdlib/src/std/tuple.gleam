import std/list

pub fn new(a, b) {
  {a, b}
}

pub fn first(tup) {
  let {a, _} = tup
  a
}

pub fn second(tup) {
  let {_, a} = tup
  a
}

pub fn swap(tup) {
  let {a, b} = tup
  {b, a}
}

pub fn fetch(haystack, needle) {
  list:find(haystack, fn(tuple) {
    case first(tuple) == needle {
    | True -> tuple |> second |> Ok
    | False -> Error([])
    }
  })
}
