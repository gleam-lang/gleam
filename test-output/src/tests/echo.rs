use std::{env, io::Read, process::Stdio};

use camino::{Utf8Path, Utf8PathBuf};
use gleam_core::{
    build::{Runtime, Target},
    io::Command,
    paths::ProjectPaths,
};

use gleam_cli::{
    fs,
    run::{self, Which},
};

fn run_and_produce_pretty_snapshot(
    target: Option<Target>,
    runtime: Option<Runtime>,
    project_directory: Utf8PathBuf,
) -> String {
    let project_root = fs::get_project_root(project_directory).expect("project root");
    let paths = ProjectPaths::new(project_root);

    let output = run_and_capture_output(&paths, "main", target, runtime)
        // Since the echo output's contains a path we will replace the `\` with a `/`
        // so that the snapshot doesn't fail on Windows in CI.
        .replace("src\\", "src/");

    let main_module_content =
        fs::read(paths.src_directory().join("main.gleam")).expect("read main module");

    format!(
        "--- main.gleam ----------------------
{main_module_content}

--- gleam run output ----------------
{output}
"
    )
}

fn run_and_capture_output(
    paths: &ProjectPaths,
    main_module: &str,
    target: Option<Target>,
    runtime: Option<Runtime>,
) -> String {
    fs::delete_directory(&paths.build_directory()).expect("delete build directory content");

    let Command {
        program,
        args,
        env,
        cwd: _,
        stdio: _,
    } = run::setup(
        paths,
        vec![],
        target,
        runtime,
        Some(main_module.into()),
        Which::Src,
        true,
    )
    .expect("run setup");

    let mut process = std::process::Command::new(&program)
        .args(args)
        .stdin(Stdio::null())
        .stdout(Stdio::null())
        .stderr(Stdio::piped())
        .envs(env.iter().map(|pair| (&pair.0, &pair.1)))
        .current_dir(paths.root())
        .spawn()
        .unwrap_or_else(|e| panic!("Failed to spawn process '{}': {}", &program, &e));

    let mut stderr = process.stderr.take().expect("take stderr");
    let mut output = String::new();
    let _ = stderr.read_to_string(&mut output).expect("read stderr");
    let _ = process.wait().expect("run with no errors");
    output
}

macro_rules! assert_echo {
    ($project_name: expr) => {
        let snapshot_name = snapshot_name(None, None, $project_name);
        insta::allow_duplicates! {
            assert_echo!(&snapshot_name, Some(Target::Erlang), None, $project_name);
            assert_echo!(&snapshot_name, Some(Target::JavaScript), Some(Runtime::Bun), $project_name);
            assert_echo!(&snapshot_name, Some(Target::JavaScript), Some(Runtime::Deno), $project_name);
            assert_echo!(&snapshot_name, Some(Target::JavaScript), Some(Runtime::NodeJs), $project_name);
        }
    };

    ($target: expr, $project_name: expr) => {
        let snapshot_name = snapshot_name(Some($target), None, $project_name);
        match $target {
            Target::JavaScript => insta::allow_duplicates! {
                assert_echo!(&snapshot_name, Some($target), Some(Runtime::Bun), $project_name);
                assert_echo!(&snapshot_name, Some($target), Some(Runtime::Deno), $project_name);
                assert_echo!(&snapshot_name, Some($target), Some(Runtime::NodeJs), $project_name);
            },
            Target::Erlang => {
                assert_echo!(&snapshot_name, Some($target), None, $project_name);
            }
        }
    };

    ($snapshot_name: expr, $target: expr, $runtime: expr, $project_name: expr) => {
        let path = fs::canonicalise(&Utf8Path::new("../test-output/cases").join($project_name))
            .expect("canonicalise path");
        let output = run_and_produce_pretty_snapshot($target, $runtime, path);
        insta::assert_snapshot!($snapshot_name.to_string(), output);
    };
}

fn snapshot_name(target: Option<Target>, runtime: Option<Runtime>, suffix: &str) -> String {
    let prefix = match (target, runtime) {
        (None, None) => "".into(),
        (None, Some(runtime)) => format!("{runtime}-"),
        (Some(target), None) => format!("{target}-"),
        (Some(target), Some(runtime)) => format!("{target}-{runtime}-"),
    };
    format!("{prefix}{suffix}")
}

#[test]
fn echo_bitarray() {
    assert_echo!(Target::JavaScript, "echo_bitarray");
    assert_echo!(Target::Erlang, "echo_bitarray");
}

#[test]
fn echo_bool() {
    assert_echo!("echo_bool");
}

#[test]
fn echo_charlist() {
    assert_echo!("echo_charlist");
}

#[test]
fn echo_custom_type() {
    assert_echo!(Target::Erlang, "echo_custom_type");
    assert_echo!(Target::JavaScript, "echo_custom_type");
}

#[test]
fn echo_dict() {
    assert_echo!("echo_dict");
}

#[test]
fn echo_float() {
    assert_echo!(Target::Erlang, "echo_float");
    assert_echo!(Target::JavaScript, "echo_float");
}

#[test]
fn echo_function() {
    assert_echo!("echo_function");
}

#[test]
fn echo_importing_module_named_inspect() {
    assert_echo!("echo_importing_module_named_inspect");
}

#[test]
fn echo_int() {
    assert_echo!("echo_int");
}

#[test]
fn echo_list() {
    assert_echo!("echo_list");
}

#[test]
fn echo_nil() {
    assert_echo!("echo_nil");
}

#[test]
fn echo_string() {
    assert_echo!("echo_string");
}

#[test]
fn echo_tuple() {
    assert_echo!("echo_tuple");
}

#[test]
fn echo_non_record_atom_tag() {
    assert_echo!(Target::Erlang, "echo_non_record_atom_tag");
}

#[test]
fn echo_circular_reference() {
    assert_echo!(Target::JavaScript, "echo_circular_reference");
}

#[test]
fn echo_singleton() {
    assert_echo!("echo_singleton");
}
