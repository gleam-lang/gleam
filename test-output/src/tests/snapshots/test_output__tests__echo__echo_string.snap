---
source: test-output/src/tests/echo.rs
expression: output
---
--- main.gleam ----------------------
pub fn main() {
  echo "hello world"
  echo "multiline
string"
  echo "string with\t\r\nescaped chars"
}


--- gleam run output ----------------
[90msrc/main.gleam:2[39m
"hello world"
[90msrc/main.gleam:3[39m
"multiline\nstring"
[90msrc/main.gleam:5[39m
"string with\t\r\nescaped chars"
